compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1364,7 +1364,7 @@ class Typer extends Namer
             tree.selector.removeAttachment(desugar.CheckIrrefutable) match {
               case Some(checkMode) =>
                 val isPatDef = checkMode == desugar.MatchCheck.IrrefutablePatDef
-                if (!checkIrrefutable(pat, sel.tpe, isPatDef) && ctx.settings.migration.value)
+                if (!checkIrrefutable(pat, sel.tpe, isPatDef) && sourceVersion == `3.1-migration`)
```

Comment:
Do we need this for `3.0-migration`?

Reply:
I think we dont. According to the [Dotty documentation](https://dotty.epfl.ch/docs/reference/changed-features/pattern-bindings.html), the type checking rules of pattern bindings will change in Dotty 3.1, not Dotty 3.0.

The following code compiles and runs successfully in 0.26-M1

```scala
val list: List[Int] = List(1)
val head :: _ = list

val listOpt: List[Option[Int]] = List(Some(1), None)
for (Some(value) <- listOpt)
  println(value)
```